### DISCUSSION TITLE
Problems with using Nextcloud OAuth trough Generic OAuth connector (ID Token not present in token response)

### DISCUSSION THREAD

**Discussion by dmigis** [requarks/wiki#3434] [Help / Questions]
**Question**
I'm struggling to make wikijs working with Nextcloud as OAuth provider via Generic OpenID/OAuth connector. After successful redirection to Nextcloud login page and granting access Wikijs throws me an "ID Token not present in token response" error. In wikijs logs there is no error, nginx and nextcloud logs show, that token was received, but user data wasn't requested. Other services with Nextcloud as OAuth provider (Jupyterhub, Vouch proxy and so on) working okay. Is it a bug or I'm doing something wrong?

![2020-09-08_10-23](https://user-images.githubusercontent.com/58209433/92447262-aeaf6800-f1bf-11ea-8dce-66b3be59228c.png)

**Host Info (please complete the following information):**
OS: Docker
Wiki.js version: 2.5.126
Database engine: Postgres 11

**Comment by tionis** [comment]
I have the same problem when trying to use Gitea as oAuth provider. Same config works fine for other applications using it as oAuth provider.

**Comment by maznu** [comment]
Same problems here with our own home-grown OIDC auth endpoint.

I can see wiki.js making a POST to our token endpoint, which responds something like this:

`{"access_token": "blahblahblah", "token_type": "Bearer", "scope": "openid openid profile email", "refresh_token": "blahblahblah", "expires_in": 3600}`

But it doesn't get as far as making a request to our userinfo endpoint.

**Comment by Doctacosa** [comment]
I'm getting the same results.

OS: Docker
Wiki.js version: 2.5.170
Database engine: Postgres 11

My homegrown endpoint appears to be working fine, as I'm able to obtain the proper tokens by using a third-party tool like Postman, but Wiki.js fails with the error described above. I can't think of any alternative to make this work.

**Comment by cse342** [comment]
I am also affected by this issue.

OS: Ubuntu Server 20.04.1 LTS
Database: Postgres 12
Wiki.js: 2.5.170

I tried it with Nextcloud and with a Forum based on Woltlab Burning Board with an Addon called "Oauth 2.0 Server".
In both cases I receive the message "ID Token not present in token response" upon granting access on the openid server.
If I checked correctly, the underlying library is passport-openidconnect, the file in charge is lib/strategy.js. If I looked it up correctly, this library hasn't received an update in nearly 4 years.

**Comment by daniel-illi** [comment]
I have the same problem.

I think wiki.js implements the OAuth 'implicit flow' to get the access token: the auth server returns the token immediately.
But all clients I used so far implement the 'authorization code' flow where the auth server returns an authorization code, which the client in turn uses to request the token from the auth server.

**Comment by Doctacosa** [comment]
> I think wiki.js implements the OAuth 'implicit flow' to get the access token: the auth server returns the token immediately.

That's an interesting idea, I wasn't aware of the difference. Checking through the documentation of my PHP OAuth library (league/oauth2-server), the implicit workflow would have the client send a `response_type` of **token**, while the authorization flow would use **code**. Testing on Wiki.js, it sends **code**, implying that it's the authorization workflow being used.

I don't pretend to be super familiar with OAuth, so if someone else could chime in on this, it would be nice!

**Comment by daniel-illi** [comment]
After having another look, I realize that I was mistaken.

The error message reads "**ID token** not present in token response".
The wiki.js OAuth module expects the `token` endpoint response to include an [JWT id token](https://openid.net/specs/openid-connect-core-1_0.html#CodeIDToken).

But when I inspect the browser requests in the developer console, I see that the redirect from the `authorize` endpoint triggers this error even before querying the `token` endpoint.
From my understanding, wiki.js should read the code from the redirect, call the `token` endpoint with the code and receive the access token and maybe an id token.

So something seems to be wired incorrectly.

**Comment by assada** [comment]
Same issue with oauth2 login. Looks like wiki.js support only OpenID type of OAuth2 protocol. =(

**Comment by troedel2546** [comment]
Same for me here with trying to connect to Mattermost as OAuth provider.

**Comment by mariosgit** [comment]
Just run into the same problem, the callback to https://wikijsthing/login/71ece951-xxx/callback?state=something&code=somestuff crashes with 500 server error. I do not see any token call, I guess this should be done under the hood.

**Comment by martinkoutecky** [comment]
Hi -- I saw the [issue](https://github.com/Requarks/wiki/issues/2392) as closed, but the current release (2.5.197) doesn't fix it for me -- should it? Anything I can do?

**Reply by Doctacosa** [reply]
Same, I just tried the new version and it doesn't change anything to the situation. I was hoping that the mention of _"Login redirect"_ in the bugs fixed would be relevant, but apparently not.

Either we're all missing something, and the Wiki.js documentation should mention the working authentication flow type, or the bug report shouldn't have been closed in the first place.

**Reply by NGPixel** [reply]
@Doctacosa It wasn't closed. It was simply converted to a discussion.

This issue has not been addressed yet.

**Reply by Doctacosa** [reply]
The GitHub email notification said that the issue was closed, then moved to discussion. If you're telling me that it doesn't remove this point from consideration, then that's fine. I understand this not being a top issue.

**Reply by martinkoutecky** [reply]
I guess we are waiting for this PR to get reviewed and merged? https://github.com/Requarks/wiki/pull/3094
